client/asset/btc/btc.go

```diff
@@ -2365,6 +2377,209 @@ func (btc *baseWallet) fund(keep uint64, minConfs uint32, lockUnspents bool,
+		reorderedFundingUTXOs := make([][]*compositeUTXO, 0, len(fundingUTXOs))
+		for _, value := range values {
+			for i, orderedValue := range orderedValues {
+				if value == orderedValue {
+					reorderedFundingUTXOs = append(reorderedFundingUTXOs, fundingUTXOs[i])
+					break
+				}
+			}
+		}
+		return reorderedFundingUTXOs
+	}
+
+	returnValues := func(allFundingUTXOs [][]*compositeUTXO) (coins []asset.Coins, redeemScripts [][]dex.Bytes, fundingCoins map[outPoint]*utxo, spents []*output, err error) {
+		coins = make([]asset.Coins, len(allFundingUTXOs))
+		fundingCoins = make(map[outPoint]*utxo)
+		spents = make([]*output, 0, len(allFundingUTXOs))
+		redeemScripts = make([][]dex.Bytes, len(allFundingUTXOs))
+		for i, fundingUTXOs := range allFundingUTXOs {
+			coins[i] = make(asset.Coins, len(fundingUTXOs))
+			redeemScripts[i] = make([]dex.Bytes, len(fundingUTXOs))
+			for j, output := range fundingUTXOs {
+				coins[i][j] = newOutput(output.txHash, output.vout, output.amount)
+				fundingCoins[outPoint{txHash: *output.txHash, vout: output.vout}] = &utxo{
+					txHash:  output.txHash,
+					vout:    output.vout,
+					amount:  output.amount,
+					address: output.address,
+				}
+				spents = append(spents, newOutput(output.txHash, output.vout, output.amount))
+				redeemScripts[i][j] = output.redeemScript
+			}
+		}
+		return
+	}
+
+	valuesCopy := make([]*asset.MultiOrderValue, len(values))
+	copy(valuesCopy, values)
+
+	// Try to fund in increasing order of value first.
+	sort.Slice(valuesCopy, func(i, j int) bool {
+		return valuesCopy[i].Value < valuesCopy[j].Value
+	})
+	allFundingUTXOs := fundInOrder(valuesCopy)
+	if len(allFundingUTXOs) == len(values) { // successfully funded all orders
+		reorderedFundingUTXOs := reorderFundingUTXOs(allFundingUTXOs, valuesCopy)
+		return returnValues(reorderedFundingUTXOs)
+	}
+
+	// If could not fully fund, fund as much as possible in the priority
+	// order.
+	allFundingUTXOs = fundInOrder(values)
+	return returnValues(allFundingUTXOs)
+}
+
+// fundMultiWithSplit creates a split transaction with one output for each
+// order in order to fund the orders. An error is returned if there is not
+// enough funds in the wallet to cover the split transaction for each order.
+func (btc *baseWallet) fundMultiWithSplit(keep uint64, values []*asset.MultiOrderValue, splitTxFeeRate, maxFeeRate, splitBuffer uint64) ([]asset.Coins, [][]dex.Bytes, uint64, error) {
+	var swapInputSize uint64
+	if btc.segwit {
+		swapInputSize = dexbtc.RedeemP2WPKHInputTotalSize
+	} else {
+		swapInputSize = dexbtc.RedeemP2PKHInputSize
+	}
+
+	var totalOutputRequired uint64
+	requiredForOrders := make([]uint64, len(values))
+	for i, value := range values {
+		req := calc.RequiredOrderFundsAlt(value.Value, swapInputSize, value.MaxSwapCount, btc.initTxSizeBase, btc.initTxSize, maxFeeRate)
+		req = req * (100 + splitBuffer) / 100
+		requiredForOrders[i] = req
+		totalOutputRequired += req
+	}
```

Comment:
Would it be useful to check existing utxos to see if we already have one that satisfies our needs here? Bots will likely cancel and replace a lot of fully unmatched orders, If these orders change by a rate that keeps the old funding coins withing the split buffer, we can reuse the utxos. If we can maximize our reuse of those outputs and minimize the number of inputs that go into the split tx (by reducing the total funding needed), we could potentially see significant fee savings.

Reply:
`fundMultiBestEffort` handles this. It tries to fund all of the orders with existing UTXOs. If it is not able to, only then is the split done.

Reply:
Example: Ignoring fees, orders of (10, 10, 10), utxos of (10, 10, 15), `maxLock = 30`, `allowSplits = true`, `fundMultiBestEffort` would fail because of `maxLock`. `fundMultiWithSplit` would consume all three inputs to create outputs of (10, 10, 10,  5), even though we already had two outputs of 10. The ideal split would just consume the utxo of value 15 and create outputs at (10, 5). Inputs are much more expensive than outputs. So the current algo almost triples the fees associated with the split transaction compared to the ideal.

Reply:
Ah good point. I'll combine the best effort logic with splitting.

Reply:
I took a swing at a hybrid approach [here](https://github.com/buck54321/dcrdex/compare/e2d6f5b...buck54321:dcrdex:fundmulti), but the diff is kinda messy, so might be easier to look at [this](https://github.com/buck54321/dcrdex/blob/4b01c6af4593fe4ee1b235d28746e2b1fffa4fde/client/asset/btc/btc.go#L2436) first. (Not tested and tests not updated)

The primary differences
- Separates available funding (`maxLock`) from allowable overfunding (`overfundBuffer`)
- Only creates splits when existing outputs don't work

A notable consequence
- Orders can be funded out of order in partial results. For instance, if we're funding values (1, 3, 1), and we have utxos (1, 1, 50), low overfund allowance, and `allowSplit = false`, the first and last orders will be funded, but not the middle, since that would require an high overlock